tiledb/cloud/dag/dag.py

```diff
@@ -792,3 +763,100 @@ def end_results_by_name(self):
             results[node.name] = node.result()
 
         return results
+
+
+class _ReplaceWith(Exception):
+    """Indicates that the object should be replaced rather than descending."""
+
+    def __init__(self, value: Any):
+        self.value = value
+
+
+def replace_stored_params(tree, loader: stored_params.ParamLoader) -> Any:
+    """Descends into data structures and replaces Stored Params with results.
+
+    This is only useful in the UDF execution environment.
+
+    Guaranteed to produce new output structures and not mutate the input.
+    Supports cyclic data structures.
+    """
+
+    def visit(arg):
+        if isinstance(arg, stored_params.StoredParam):
+            raise _ReplaceWith(loader.load(arg))
+
+    return _ResultReplacer(visit)._replace_internal(tree)
+
+
+def _replace_nodes_with_results(tree):
+    """Descends into data structures and replaces Node IDs with their values.
+
+    Offers the same guarantees as :func:`replace_stored_params`.
+    """
+
+    def visit(arg):
+        if isinstance(arg, Node):
+            raise _ReplaceWith(arg.result())
+
+    return _ResultReplacer(visit)._replace_internal(tree)
+
+
+class _ResultReplacer:
+    """Class to visit nodes and replace specific visited nodes.
+
+    This class takes as a parameter a "visit" function.  When the visit function
+    is called, if it raises a :class:`_ReplaceWith`, instead of descending
+    further into the argument, the :class:`_ResultReplacer` will replace the
+    given argument with the value the :class:`_ReplaceWith` contains.
+    """
+
+    def __init__(self, visit: Callable[[Any], Any]):
+        self.seen: Dict[int, Any] = {}
+        self.visit = visit
+
+    def _replace_internal(self, arg):
+        if isinstance(arg, (str, bytes)):
+            # Special-case these since they're weird sequences.
+            return arg
+
+        original_id = id(arg)
+        try:
+            return self.seen[original_id]
+        except KeyError:
+            pass  # We haven't seen this object yet; continue.
+
+        # First, handle if this is something to replace directly.
```

Comment:
Per the top-level comment, _at very least_ I would like to have an inline comment right here that says something like "FYI, think of `_ReplaceWith` + `try` here as a funny-looking `Optional(arg)` that is unwrapped in the `except` clause."

Reply:
Decided to use a class-based approach. The reason I did it this way was to avoid ambiguity with `None`, but I think my new approach is better.

Reply:
👍